handler/oauth2/flow_refresh.go

```diff
@@ -90,8 +94,27 @@ func (c *RefreshTokenGrantHandler) HandleTokenEndpointRequest(ctx context.Contex
 	request.SetRequestedScopes(originalRequest.GetRequestedScopes())
 	request.SetRequestedAudience(originalRequest.GetRequestedAudience())
 
+	var delegatingClient fosite.Client = nil
+	if originalRequest.GetDelegatingClient() != nil {
+		// original request was token exchange
+		// reload client from storage to ensure that may_act is up to date in case of eventual revocation
+		delegatingClient, err = c.Store.GetClient(ctx, originalRequest.GetDelegatingClient().GetID())
+		if err != nil {
+			return errors.WithStack(fosite.ErrInvalidClient.WithHint("The delegating OAuth2 Client does not exist."))
+		}
+		// keep delegating client for following refresh requests
+		request.SetDelegatingClient(delegatingClient)
+
+		// check if delegating client (still) has may_act claim for current client
+		if !delegatingClient.GetMayAct().HasOneOf(request.GetClient().GetID()) {
+			return errors.WithStack(fosite.ErrUnauthorizedClient.WithHint("The OAuth 2.0 Client is not allowed to perform a token exchange for the given subject token."))
+		}
+	}
+
 	for _, scope := range originalRequest.GetGrantedScopes() {
-		if !c.ScopeStrategy(request.GetClient().GetScopes(), scope) {
+		if !c.ScopeStrategy(request.GetClient().GetScopes(), scope) &&
+			// if it is a refresh of an exchanged token, check scopes against those of the delegating client
+			(delegatingClient == nil || !c.ScopeStrategy(delegatingClient.GetScopes(), scope)) {
 			return errors.WithStack(fosite.ErrInvalidScope.WithHintf("The OAuth 2.0 Client is not allowed to request scope \"%s\".", scope))
```

Comment:
I think there may be a miss understanding.
Here we check that the granted scopes are either available on the requesting client as previously (in the case this is a refresh of an exchanged token, this would be the client that performed the token exchange), or on the scopes of the delegating client (the client that provided the token that  was exchanged).
This was an implementation decision we made in which we allow a client (performing the token exchange) to request scopes available on the delegating client or on its own client.
If there is a delegating client (implies that the refresh flow is being done for an exchanged token) the requested scopes can be declared on either of the clients.
At this point we cannot distinguish, if the requested scopes should be on the delegating client or on the own client.